### PR TITLE
Add a new resource that links a token to a policy

### PR DESCRIPTION
**Pull request by randomswdev**
This pull request implements the feature described in issue https://github.com/terraform-providers/terraform-provider-consul/issues/125.
Implemented a new resource to track the link between a token and a policy.

I also checked if it would be possible to prevent consul_acl_token from overwriting the policies if the policies attribute was not included in the resource definition. I think this is not possible without a change to the SDK.
It would be possible to know if an attribute is included or not in the resource definition using the call:
`d.get(parts, getSourceConfig|getSourceExact)`
but the method is private to the schema namespace.

Closes  https://github.com/terraform-providers/terraform-provider-consul/issues/125

### CODE REVIEW

File: consul/resource_consul_acl_token_policy_attachment_test.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	rules = "node \"\" { policy = \"read\" }"
+	datacenters = [ "dc1" ]
+}
+
+resource "consul_acl_token" "test" {
+	description = "test"
+	local = true
+
+	lifecycle {
+		ignore_ttributes = ["policies"]
```
Comment (greut) [inline consul/resource_consul_acl_token_policy_attachment_test.go]: typo?

Comment (randomswdev) [reply consul/resource_consul_acl_token_policy_attachment_test.go]: Yes, the file content was wrong. I just updated it, pushing a new revision.

**Comment by remilapeyre** [comment]
Thanks for opening this PR @randomswdev !

> I also checked if it would be possible to prevent consul_acl_token from overwriting the policies if the policies attribute was not included in the resource definition. I think this is not possible without a change to the SDK.
> It would be possible to know if an attribute is included or not in the resource definition using the call:
> `d.get(parts, getSourceConfig|getSourceExact)`
> but the method is private to the schema namespace.

This still wouldn't be enough: the `local` attribute for example is marked as `ForceNew` and a changing it would  create another token to replace the current one. There is no way to attach the policies to the new token if we don't reapply the `consul_acl_token_policy_attachment` resource too.

**Comment by randomswdev** [comment]
> This still wouldn't be enough: the `local` attribute for example is marked as `ForceNew` and a changing it would create another token to replace the current one. There is no way to attach the policies to the new token if we don't reapply the `consul_acl_token_policy_attachment` resource too.

I just checked this scenario. Changing the `local` attribute, forces a token to be recreated. When a token gets recreated, its id changes . This id is used as the value of the `token_id` attribute of the `consul_acl_token_policy_attachment` resource; because `token_id` is also set as `ForceNew`, the token id's change forces also the attachment to be recreated, thus adding again the policy to the (new) token.

So, I f I understood correctly your scenario, I think the code should work as it is.

**Comment by remilapeyre** [comment]
>>>This id is used as the value of the token_id attribute of the consul_acl_token_policy_attachment resource; because token_id is also set as ForceNew, the token id's change forces also the attachment to be recreated, thus adding again the policy to the (new) token.

Yes but this only works when the `consul_acl_token` and `consul_acl_token_policy_attachment` resources live in the same Terraform configuration, when you have all resources in the same Terraform configuration:

```hcl
resource "consul_acl_policy" "test" {
	name = "test"
	rules = "node \"\" { policy = \"read\" }"
	datacenters = [ "dc1" ]
}
resource "consul_acl_token" "test" {
	description = "test"
	lifecycle {
		ignore_changes = ["policies"]
	}
}
resource "consul_acl_token_policy_attachment" "test" {
    token_id = "${consul_acl_token.test.id}"
    policy = "${consul_acl_policy.test.name}"
}
```
it works well but if you have the token and the policy in separate configuration, you cannot use a reference but have to hardcoded the token as of now :
```hcl
resource "consul_acl_token" "test" {
	description = "test"
	local = true
	lifecycle {
		ignore_changes = ["policies"]
	}
}
```
with in another folder:
```hcl
resource "consul_acl_policy" "test" {
	name = "test"
	rules = "node \"\" { policy = \"read\" }"
	datacenters = [ "dc1" ]
}
resource "consul_acl_token_policy_attachment" "test" {
    token_id = "c597fb2c-6098-474f-9f03-dead07ae2068"
    policy = "${consul_acl_policy.test.name}"
}
```
In this case after changing the `local` parameter of `consul_acl_token.test`, updating the `consul_acl_token_policy_attachment.test` will break as a new UUID will have been generated. I think we could fix this by adding a new `accessor_id` in the `consul_acl_token` resource so the user can choose a permanent UUID (or use https://www.terraform.io/docs/providers/random/r/uuid.html to create one). Does this make sense?

**Comment by randomswdev** [comment]
Added the `accessor_id` to the consul_acl_token resource and updated the relevant documentation.

**Comment by remilapeyre** [comment]
Hi @randomswdev, thanks for adding this, I will make a final review today and will merge this :)

**Comment by remilapeyre** [comment]
Thanks for the help on this @randomswdev 👍 !

**Pull request merged**